x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/local/PushDownVectorSimilarityFunctions.java

```diff
@@ -81,57 +83,26 @@ protected LogicalPlan rule(LogicalPlan plan, LocalLogicalOptimizerContext contex
     }
 
     private static Expression replaceFieldsForFieldTransformations(
-        VectorSimilarityFunction similarityFunction,
+        Expression e,
         Map<Attribute.IdIgnoringWrapper, Attribute> addedAttrs,
-        LocalLogicalOptimizerContext context
+        BlockLoaderExpression.Fuse fuse
     ) {
-        // Only replace if exactly one side is a literal and the other a field attribute
-        if ((similarityFunction.left() instanceof Literal ^ similarityFunction.right() instanceof Literal) == false) {
```

Comment:
Nice! It's much better to let the Expression deal with the details and make this generic 👍